operator/pkg/manifest/installer.go

```diff
@@ -267,19 +268,48 @@ func parseKubectlVersion(kubectlStdout string) (*goversion.Version, *goversion.V
 }
 
 // ApplyAll applies all given manifests using kubectl client.
-func ApplyAll(manifests name.ManifestMap, version pkgversion.Version, iop *v1alpha1.IstioOperatorSpec, opts *kubectlcmd.Options) (CompositeOutput, error) {
+func ApplyAll(manifests name.ManifestMap, version pkgversion.Version, iop *v1alpha1.IstioOperatorSpec,
+	opts *kubectlcmd.Options, certDir, clusterID string) (CompositeOutput, error) {
 	scope.Infof("Preparing manifests for these components:")
 	for c := range manifests {
 		scope.Infof("- %s", c)
 	}
 	scope.Infof("Component dependencies tree: \n%s", installTreeString())
-	if _, err := InitK8SRestClient(opts.Kubeconfig, opts.Context); err != nil {
+	restConfig, err := InitK8SRestClient(opts.Kubeconfig, opts.Context)
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
 		return nil, err
 	}
 	// Set up the namespace for installation
 	if err := createNamespace(iop.Namespace); err != nil {
 		return nil, err
 	}
+
+	if clusterID == "" {
+		n, err := client.CoreV1().Namespaces().Get(context2.TODO(), "kube-system", metav1.GetOptions{})
+		if err != nil {
+			scope.Warnf("failed to get uid for namespace %s to use as clusterID, will use `default` as "+
+				"clusterID", "kube-system")
+			clusterID = "default"
+		} else {
+			clusterID = string(n.UID)
+		}
+	}
+
+	//try to plugIn external certs
+	if certDir != "" {
+		err := cert.PlugInProvisionedCert(client, certDir, iop.Namespace, opts.DryRun)
+		if err != nil {
+			scope.Infof("plugin specified cert and key failed, will generate self signed root cert and two intermediate ca certs")
+		}
+		err = cert.GenerateCertsAndKey(client, iop.Namespace, opts.DryRun, certDir, clusterID)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Doesn't `GenerateCertsAndKey` need to be after line 306 within the error path? Otherwise it will always overwrite the plugged cert.

Reply:
there is a condition check in `GenerateCertsAndKey` to make sure we generate key and certs only if the related `cacert` secret does not exist so the owerwritting will not happen.
Agree we can move this  after line 306 to make it more efficient and clear